### PR TITLE
Implement message deletion

### LINKED ISSUE #70: feature request: relay deleted messages
it would be great if the bridge relayed deleted/redacted messages. This would probably require a mapping of matrix eventId -> slack ts.

### LINKED ISSUE #100: Add support for matrix eventid to slack ts mapping
If we can implement a bidirectional eventid <> ts store, we can do a whole bunch of fun things like:

* Bridge redactions <> deletions. (#70 )
* Support replies <> threads bridging.
* Convert edits into replies to have more context on the matrix side.
* Bridge pinned messages.

### PR DESCRIPTION
**Pull request by Cadair**
fixes #70 and fixes #100 (although not all the other things that this enables)

### CODE REVIEW

File: lib/BaseSlackHandler.js [hunk 1]
```diff
@@ -26,6 +26,10 @@ function BaseSlackHandler(main) {
 BaseSlackHandler.prototype.replaceChannelIdsWithNames = function(message, token) {
     var main = this._main;
 
+    if (message.text === undefined) {
```
Comment (Half-Shot) [inline lib/BaseSlackHandler.js]: God I wish for async support.

File: lib/BridgedRoom.js [hunk 2]
```diff
@@ -188,6 +190,38 @@ BridgedRoom.prototype.toEntry = function() {
...
+    const eventStore = this._main.getEventStore();
+    const event = await eventStore.getEntryByMatrixId(message.room_id, message.redacts);
+
+    // If we don't get an event then exit
+    if (event === null) {
+        log.debug("Could not find event to delete.");
+        return Promise.resolve();
+    }
+
+    var body = {channel: this._slack_channel_id,
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: VAR!?!?!

File: lib/BridgedRoom.js [hunk 3]
```diff
@@ -188,6 +190,38 @@ BridgedRoom.prototype.toEntry = function() {
     return entry;
 };
 
+BridgedRoom.prototype.onMatrixRedaction = async function(message) {
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: wait we do have await? Neat

File: lib/BaseSlackHandler.js (hunk 1 above)
Comment (Half-Shot) [reply lib/BaseSlackHandler.js]: as per my comment below, make the function async and just return message :)

File: lib/BridgedRoom.js [hunk 4]
```diff
@@ -234,35 +273,45 @@ BridgedRoom.prototype.onMatrixMessage = function(message) {
 BridgedRoom.prototype.onSlackMessage = function(message) {
     return this._main.getGhostForSlackMessage(message).then((ghost) => {
         return ghost.update(message, this).then(() => {
-            return this._handleSlackMessage(message, ghost);
+            return this._handleSlackMessage(message, ghost).catch((err) => {
+                log.warn("Failed to process event");
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: Do we want to log 'err'?

File: lib/BridgedRoom.js [hunk 5]
```diff
@@ -234,35 +273,45 @@ BridgedRoom.prototype.onMatrixMessage = function(message) {
 BridgedRoom.prototype.onSlackMessage = function(message) {
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: Surely this is a great candidate for async?

File: lib/BridgedRoom.js [hunk 6]
```diff
@@ -234,35 +273,45 @@ BridgedRoom.prototype.onMatrixMessage = function(message) {
...
     return this._main.getGhostForSlackMessage(message).then((ghost) => {
         return ghost.update(message, this).then(() => {
-            return this._handleSlackMessage(message, ghost);
+            return this._handleSlackMessage(message, ghost).catch((err) => {
+                log.warn("Failed to process event");
+            });
         });
+    }).catch((err) => {
+        log.warn("Failed to process event");
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: Same as before, log `err`?

File: lib/SlackGhost.js [hunk 7]
```diff
@@ -160,7 +162,7 @@ SlackGhost.prototype.updateAvatar = function(message, room) {
     });
 };
 
-SlackGhost.prototype.sendText = function(room_id, text) {
+SlackGhost.prototype.sendText = function(room_id, text, slackRoomID, slackEventTS) {
```
Comment (Half-Shot) [inline lib/SlackGhost.js]: Question, what's with the roomId and timestamps?

Comment (Cadair) [reply lib/SlackGhost.js]: My insane variable naming conventions?

**Comment by Half-Shot** [comment]
Looks good, happy to merge.

**Comment by Half-Shot** [comment]
@Cadair feck. Could you sign off for me :)

**Comment by Cadair** [comment]
Signed off By Stuart Mumford (stuart@cadair.com)

**Pull request merged**